examples/gym/solve_gridworld.py

```diff
@@ -0,0 +1,201 @@
+import logging
+import time
+import sys
+from typing import Tuple
+
+import numpy as np
+from sqlalchemy import create_engine
+
+from rlmolecule.tree_search.reward import Reward
+from rlmolecule.alphazero.tfalphazero_problem import TFAlphaZeroProblem
+from rlmolecule.gym.gym_problem import GymProblem
+from rlmolecule.gym.gym_state import GymEnvState
+from rlmolecule.gym.alphazero_gym import AlphaZeroGymEnv
+
+from examples.gym.tf_model import policy_model_2
+import examples.gym.gridworld_env as gw
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+# NOTE: These class definitions need to stay outside of construct_problem
+# or you will error out on not being able to pickle/serialize them.
+class GridWorldEnv(AlphaZeroGymEnv):
+    def __init__(self,
+                 configured_env: AlphaZeroGymEnv,
+                 **kwargs):
+        super().__init__(configured_env, **kwargs)
+
+    def reset(self):
+        return self.env.reset()
+
+    def get_obs(self) -> np.ndarray:
+        return self.env.get_obs()
+
+
+class GridWorldProblem(GymProblem, TFAlphaZeroProblem):
+    def __init__(self, *,
+                 env: gw.GridEnv,
+                 engine: "sqlalchemy.engine.Engine",
+                 reward_class: Reward,
+                 **kwargs) -> None:
+        super().__init__(reward_class=reward_class, engine=engine, env=env, **kwargs)
+
+    def policy_model(self) -> "tf.keras.Model":
+        obs_shape = self.env.reset().shape
+        return policy_model_2(obs_dim=obs_shape,
+                              hidden_layers=2,
+                              conv_layers=1,
+                              filters_dim=[16],
+                              kernel_dim=[2],
+                              strides_dim=[1],
+                              hidden_dim=64)
+
+    def get_policy_inputs(self, state: GymEnvState) -> dict:
+        return {
+            "obs": state.env.get_obs(),
+            "steps": 0.*np.array([np.float64(self.env.episode_steps)])
+        }
+
+    def get_reward(self, state: GymEnvState) -> Tuple[float, dict]:
+        return state.cumulative_reward, {}
```

Comment:
Interesting... any reason this couldn't get moved up to GymProblem?

Reply:
another good catch.  if we stick to the definition of env.step as always returning the step / terminal reward (which is what env.step is supposed to do) then yes ... i didn't initially migrate the hallway env properly to fit this definition.  but done now!